render-ws-java-client/src/main/java/org/janelia/render/client/newsolver/StackAlignmentErrorClient.java

```diff
@@ -122,27 +158,95 @@ private static AlignmentErrors computeSolveItemErrors(final ResolvedTileSpecsWit
-
-			final double vDiff = WorkerTools.computeAlignmentError(
-					crossLayerModel,
-					stitchingModel,
-					pTileSpec,
-					qTileSpec,
-					pTileSpec.getLastTransform().getNewInstance(),
-					qTileSpec.getLastTransform().getNewInstance(),
-					match.getMatches());
+			final double errorValue;
+
+			if (calculateResiduals) {
+				errorValue = deriveRootMeanSquaredError(stackId,
+														matchCollectionId,
+														match,
+														pTileSpec,
+														qTileSpec,
+														tileIdToMatchTileSpec);
+			} else {
+				errorValue = WorkerTools.computeAlignmentError(crossLayerModel,
+															   stitchingModel,
+															   pTileSpec,
+															   qTileSpec,
+															   pTileSpec.getLastTransform().getNewInstance(),
+															   qTileSpec.getLastTransform().getNewInstance(),
+															   match.getMatches());
+			}
 
 			final OrderedCanvasIdPair pair = match.toOrderedPair();
-			alignmentErrors.addError(pair, vDiff);
+			alignmentErrors.addError(pair, errorValue);
 		}
 
 		LOG.info("computeSolveItemErrors, exit");
+
 		return alignmentErrors;
 	}
 
+	private static double deriveRootMeanSquaredError(final StackId stackId,
+													 final MatchCollectionId matchCollectionId,
+													 final CanvasMatches match,
+													 final TileSpec pAlignedTileSpec,
+													 final TileSpec qAlignedTileSpec,
+													 final Map<String, TileSpec> tileIdToMatchTileSpec)
+			throws NoninvertibleModelException {
+
+		final String pTileId = pAlignedTileSpec.getTileId();
+		final String qTileId = qAlignedTileSpec.getTileId();
+		final ResidualCalculator residualCalculator = new ResidualCalculator();
+		final ResidualCalculator.InputData inputData = new ResidualCalculator.InputData(pTileId,
+																						qTileId,
+																						stackId,
+																						matchCollectionId,
+																						false);
+		final List<PointMatch> worldMatchList = match.getMatches().createPointMatches();
+
+		final TileSpec pMatchTileSpec = getMatchTileSpec(pAlignedTileSpec, tileIdToMatchTileSpec);
+		final TileSpec qMatchTileSpec = getMatchTileSpec(qAlignedTileSpec, tileIdToMatchTileSpec);
+		final List<PointMatch> localMatchList = ResidualCalculator.convertMatchesToLocal(worldMatchList,
+																						 pMatchTileSpec,
+																						 qMatchTileSpec);
+
+		if (localMatchList.isEmpty()) {
+			throw new IllegalArgumentException(inputData.getMatchCollectionId() + " has " +
+											   worldMatchList.size() + " matches between " + pTileId + " and " +
+											   qTileId + " but none of them are invertible");
+		}
+
+		final ResidualCalculator.Result result = residualCalculator.run(stackId,
+																		inputData,
+																		localMatchList,
+																		pAlignedTileSpec,
+																		qAlignedTileSpec);
+
+		return result.getRootMeanSquareError();
+	}
+
+	private static TileSpec getMatchTileSpec(final TileSpec alignedTileSpec,
```

Comment:
Would `getUntransformedTileSpec` be a better name for this method?
Also, it could be precomputed for all nodes in `tilesAndMatches.getTiles()` before looping over all matches. This would probably be more reusable and the way it's done right now introduces some mental clutter in form of a mutable map.

Reply:
With regard  to the method name ... I don't think `getUntransformedTileSpec` is accurate because the tile spec may still be transformed for lens correction.  I'll switch this to `buildTileSpecUsedForMatchDerivation` and hope that is clearer.  As you suggested, I now build the tileIdToMatchTileSpec map up front rather than as we go.